File: solutions/2026-08-05/python/2026-08-05-remove-methods-from-project.py
```python
# Problem: Remove Methods From Project
# Link: https://leetcode.com/problems/remove-methods-from-project/
# Approach:
# The problem can be modeled as a directed graph where methods are nodes and invocations are edges.
# We need to identify methods that are "suspicious". A method is suspicious if it is method 'k' or
# is invoked (directly or indirectly) by method 'k'.
#
# First, we build the graph. Since we are interested in what methods invoke other methods, we can
# represent this with an adjacency list where graph[u] contains a list of methods that 'u' invokes.
# We also need to track which methods are *invoked by* other methods. This will be useful later
# to check if a suspicious group can be removed. We can use a 'reverse_graph' or a 'parent_count'
# to track this. A 'parent_count' for each method will store how many *non-suspicious* methods
# invoke it.
#
# We can use Breadth-First Search (BFS) or Depth-First Search (DFS) starting from 'k' to find all
# methods that are suspicious (reachable from 'k'). Let's call this set 'suspicious_methods'.
#
# Now, for a group of suspicious methods to be removed, no method *outside* the group can invoke
# any method *inside* the group.
#
# This means we need to identify suspicious methods that are *not* invoked by any non-suspicious method.
#
# A more efficient way to think about this is to find all methods that *can* be removed.
# A method 'm' can be removed if:
# 1. 'm' is suspicious (reachable from 'k').
# 2. No non-suspicious method invokes 'm'.
#
# Let's refine the approach:
# 1. Build the graph: Adjacency list `adj` where `adj[u]` lists methods `v` that `u` invokes.
# 2. Build the reverse graph: Adjacency list `rev_adj` where `rev_adj[v]` lists methods `u` that invoke `v`.
# 3. Find all suspicious methods: Use BFS/DFS starting from `k` to find all methods reachable from `k`. Store them in a set `suspicious_set`.
# 4. Identify methods that can be removed: Iterate through all methods from 0 to n-1.
#    A method `i` can be removed if:
#    a. `i` is in `suspicious_set`.
#    b. For every method `caller` that invokes `i` (i.e., `caller` is in `rev_adj[i]`), `caller` must *also* be suspicious.
#       This is equivalent to checking if `i` has any *non-suspicious* callers. If `i` has no callers, it can be removed.
#       If all its callers are also suspicious, it can also be removed (as they would be removed together).
#       Therefore, a suspicious method `i` can be removed if it has *no non-suspicious callers*.
#
# Let's rethink step 4 to simplify.
# We want to remove a set S of methods such that:
# 1. S contains 'k' and all methods reachable from 'k' (directly or indirectly).
# 2. No method 'm' not in S invokes any method 's' in S.
#
# This condition means that if a method 'm' is *not* suspicious, it cannot invoke any method that *is* suspicious.
#
# Let's consider the methods that are *not* suspicious. These are the methods not reachable from 'k'.
# If any non-suspicious method invokes a suspicious method, then the suspicious method cannot be removed
# (because its removal would violate condition 2).
#
# So, the logic is:
# 1. Find all suspicious methods using BFS/DFS from `k`.
# 2. For each suspicious method `s`:
#    Check all methods `caller` that invoke `s` (using the reverse graph).
#    If any `caller` is *not* in the set of suspicious methods, then `s` cannot be removed.
#    If all `caller`s of `s` are themselves suspicious, then `s` can potentially be removed.
#
# This leads to the final strategy:
# 1. Build the adjacency list `adj` and the reverse adjacency list `rev_adj`.
# 2. Perform a BFS starting from `k` to find all methods `suspicious_set` that are reachable from `k`.
# 3. Initialize a set `removable_suspicious_methods`.
# 4. Iterate through each method `m` from 0 to n-1.
#    If `m` is in `suspicious_set`:
#       Check if `m` has any *non-suspicious* callers.
#       A caller `c` is non-suspicious if `c` is *not* in `suspicious_set`.
#       If `rev_adj[m]` is empty (no callers), then `m` can be removed.
#       If all callers `c` in `rev_adj[m]` are such that `c` is *also* in `suspicious_set`, then `m` can be removed.
#       In other words, `m` can be removed if there is NO caller `c` such that `c` is NOT in `suspicious_set`.
#
# Let's refine step 4:
# We want to identify the set of methods that *remain*. A method `i` remains if it's *not* suspicious, OR
# if it *is* suspicious but cannot be removed.
# A suspicious method `s` cannot be removed if there exists a non-suspicious method `c` such that `c` invokes `s`.
#
# So, the set of methods to remove are those suspicious methods `s` for which *all* their callers are also suspicious.
#
# Final Algorithm:
# 1. Build `adj` (forward graph) and `rev_adj` (reverse graph).
# 2. Use BFS/DFS from `k` to find all `suspicious_set` (methods reachable from `k`).
# 3. Create a set `can_be_removed_set`.
# 4. For each method `m` in `suspicious_set`:
#    Assume `m` can be removed initially (`can_remove_m = True`).
#    For each `caller` in `rev_adj[m]`:
#        If `caller` is *not* in `suspicious_set`:
#            Then `m` cannot be removed. Set `can_remove_m = False` and break.
#    If `can_remove_m` is still `True` after checking all callers, add `m` to `can_be_removed_set`.
#
# 5. Construct the result: Iterate through all methods from 0 to n-1. If a method `i` is *not* in `can_be_removed_set`, add it to the result.
#
# This algorithm ensures that only methods that are part of the "suspicious subtree" and are not
# "invoked from outside" are removed. If any suspicious method is invoked by a non-suspicious method,
# then *none* of the suspicious methods that are part of that invoked path can be removed if the
# invoking method `k` itself is not removable due to being invoked by a non-suspicious method.
#
# Example 1: n = 4, k = 1, invocations = [[1,2],[0,1],[3,2]]
# adj = {0: [1], 1: [2], 3: [2]}
# rev_adj = {1: [0], 2: [1, 3]}
#
# BFS from k=1:
# Queue: [1]
# visited: {1}
# suspicious_set = {1}
#
# Pop 1: neighbors are [2]. Add 2 to queue. visited = {1, 2}. suspicious_set = {1, 2}
# Queue: [2]
#
# Pop 2: neighbors are [].
# Queue: []
#
# suspicious_set = {1, 2}
#
# Now check which suspicious methods can be removed:
# For m = 1 (in suspicious_set):
#   rev_adj[1] = [0]
#   Is caller 0 in suspicious_set? No (0 not in {1, 2}).
#   So, method 1 cannot be removed because it's invoked by non-suspicious method 0.
#
# For m = 2 (in suspicious_set):
#   rev_adj[2] = [1, 3]
#   Caller 1: Is 1 in suspicious_set? Yes.
#   Caller 3: Is 3 in suspicious_set? No (3 not in {1, 2}).
#   So, method 2 cannot be removed because it's invoked by non-suspicious method 3.
#
# can_be_removed_set = {}
#
# Result: Methods not in can_be_removed_set.
# 0: not in {} -> add 0
# 1: not in {} -> add 1
# 2: not in {} -> add 2
# 3: not in {} -> add 3
# Output: [0, 1, 2, 3] - Correct.
#
# Example 2: n = 5, k = 0, invocations = [[1,2],[0,2],[0,1],[3,4]]
# adj = {0: [2, 1], 1: [2], 3: [4]}
# rev_adj = {1: [0], 2: [0, 1], 4: [3]}
#
# BFS from k=0:
# Queue: [0]
# visited: {0}
# suspicious_set = {0}
#
# Pop 0: neighbors are [2, 1]. Add 2, 1 to queue. visited = {0, 1, 2}. suspicious_set = {0, 1, 2}
# Queue: [2, 1]
#
# Pop 2: neighbors are [].
# Queue: [1]
#
# Pop 1: neighbors are [2]. 2 is already visited.
# Queue: []
#
# suspicious_set = {0, 1, 2}
#
# Now check which suspicious methods can be removed:
# For m = 0 (in suspicious_set):
#   rev_adj[0] is empty. No callers.
#   Method 0 can be removed. Add 0 to can_be_removed_set.
#
# For m = 1 (in suspicious_set):
#   rev_adj[1] = [0]
#   Caller 0: Is 0 in suspicious_set? Yes.
#   Method 1 can be removed. Add 1 to can_be_removed_set.
#
# For m = 2 (in suspicious_set):
#   rev_adj[2] = [0, 1]
#   Caller 0: Is 0 in suspicious_set? Yes.
#   Caller 1: Is 1 in suspicious_set? Yes.
#   Method 2 can be removed. Add 2 to can_be_removed_set.
#
# can_be_removed_set = {0, 1, 2}
#
# Result: Methods not in can_be_removed_set.
# 0: in {0, 1, 2} -> skip
# 1: in {0, 1, 2} -> skip
# 2: in {0, 1, 2} -> skip
# 3: not in {0, 1, 2} -> add 3
# 4: not in {0, 1, 2} -> add 4
# Output: [3, 4] - Correct.
#
# Example 3: n = 3, k = 2, invocations = [[1,2],[0,1],[2,0]]
# adj = {1: [2], 0: [1], 2: [0]}
# rev_adj = {2: [1], 1: [0], 0: [2]}
#
# BFS from k=2:
# Queue: [2]
# visited: {2}
# suspicious_set = {2}
#
# Pop 2: neighbors are [0]. Add 0 to queue. visited = {2, 0}. suspicious_set = {2, 0}
# Queue: [0]
#
# Pop 0: neighbors are [1]. Add 1 to queue. visited = {2, 0, 1}. suspicious_set = {2, 0, 1}
# Queue: [1]
#
# Pop 1: neighbors are [2]. 2 is already visited.
# Queue: []
#
# suspicious_set = {0, 1, 2}
#
# Now check which suspicious methods can be removed:
# For m = 0 (in suspicious_set):
#   rev_adj[0] = [2]
#   Caller 2: Is 2 in suspicious_set? Yes.
#   Method 0 can be removed. Add 0 to can_be_removed_set.
#
# For m = 1 (in suspicious_set):
#   rev_adj[1] = [0]
#   Caller 0: Is 0 in suspicious_set? Yes.
#   Method 1 can be removed. Add 1 to can_be_removed_set.
#
# For m = 2 (in suspicious_set):
#   rev_adj[2] = [1]
#   Caller 1: Is 1 in suspicious_set? Yes.
#   Method 2 can be removed. Add 2 to can_be_removed_set.
#
# can_be_removed_set = {0, 1, 2}
#
# Result: Methods not in can_be_removed_set.
# 0: in {0, 1, 2} -> skip
# 1: in {0, 1, 2} -> skip
# 2: in {0, 1, 2} -> skip
# Output: [] - Correct.
#
# The condition "If it is not possible to remove all the suspicious methods, none should be removed."
# This condition is implicitly handled by our algorithm. If any suspicious method is invoked by a
# non-suspicious method, then *that* suspicious method cannot be removed. If *any* suspicious method
# cannot be removed due to this constraint, then we effectively cannot remove the entire group of
# suspicious methods.
#
# Let's re-read this carefully: "A group of methods can only be removed if no method outside the group
# invokes any methods within it."
#
# This implies that *if* we decide to remove a set of suspicious methods, that set must be closed under
# invocation from non-suspicious methods. Any suspicious method `s` that is invoked by a non-suspicious
# method `c` means `s` cannot be part of *any* removable set of suspicious methods if `c` is not also removed.
#
# The current algorithm identifies suspicious methods that, *individually*, are not invoked by non-suspicious methods.
#
# Consider the case where a set of suspicious methods forms a cycle, and one of them is invoked by a non-suspicious method.
# Example: n=4, k=0, invocations=[[0,1], [1,2], [2,0], [3,0]]
# adj = {0: [1], 1: [2], 2: [0], 3: [0]}
# rev_adj = {1: [0], 2: [1], 0: [2, 3]}
#
# BFS from k=0:
# suspicious_set = {0, 1, 2}
#
# Check removable:
# m=0: rev_adj[0] = [2, 3]. Caller 3 is NOT suspicious. So 0 cannot be removed.
# m=1: rev_adj[1] = [0]. Caller 0 IS suspicious.
# m=2: rev_adj[2] = [1]. Caller 1 IS suspicious.
#
# So, `can_be_removed_set` will be empty for methods 1 and 2 if method 0 cannot be removed.
#
# This suggests that if *any* method `s` within the `suspicious_set` is invoked by a non-suspicious method,
# then *none* of the suspicious methods can be removed.
#
# This changes the logic for "If it is not possible to remove all the suspicious methods, none should be removed."
#
# If there exists *any* method `s` in `suspicious_set` such that `rev_adj[s]` contains a caller `c` where `c` is *not* in `suspicious_set`,
# then it's impossible to remove *any* suspicious methods, and we should return all original methods.
#
# Corrected Algorithm:
# 1. Build `adj` (forward graph) and `rev_adj` (reverse graph).
# 2. Use BFS/DFS from `k` to find all `suspicious_set` (methods reachable from `k`).
# 3. Check if any suspicious method is invoked by a non-suspicious method:
#    Set `can_remove_any = True`.
#    For each method `m` in `suspicious_set`:
#        For each `caller` in `rev_adj[m]`:
#            If `caller` is *not* in `suspicious_set`:
#                `can_remove_any = False`
#                break from inner loop
#        If `not can_remove_any`:
#            break from outer loop
#
# 4. If `not can_remove_any`:
#    Return all methods from 0 to n-1.
#
# 5. If `can_remove_any` is `True`:
#    This means all methods in `suspicious_set` are only invoked by other methods within `suspicious_set`
#    (or have no callers).
#    In this case, all methods in `suspicious_set` can be removed.
#    Construct the result by including all methods `i` from 0 to n-1 where `i` is *not* in `suspicious_set`.
#
# Let's re-test with Example 1: n = 4, k = 1, invocations = [[1,2],[0,1],[3,2]]
# suspicious_set = {1, 2}
#
# Check `can_remove_any`:
# m = 1: rev_adj[1] = [0]. Caller 0 is NOT in suspicious_set.
# `can_remove_any` becomes False. Break.
#
# Since `not can_remove_any` is True, return all methods: [0, 1, 2, 3]. Correct.
#
# Let's re-test with Example 2: n = 5, k = 0, invocations = [[1,2],[0,2],[0,1],[3,4]]
# suspicious_set = {0, 1, 2}
#
# Check `can_remove_any`:
# m = 0: rev_adj[0] is empty. No non-suspicious callers.
# m = 1: rev_adj[1] = [0]. Caller 0 IS in suspicious_set.
# m = 2: rev_adj[2] = [0, 1]. Caller 0 IS in suspicious_set. Caller 1 IS in suspicious_set.
#
# No caller found outside `suspicious_set` for any method in `suspicious_set`.
# `can_remove_any` remains True.
#
# Since `can_remove_any` is True, return methods NOT in `suspicious_set`.
# Methods not in {0, 1, 2} are 3, 4.
# Output: [3, 4]. Correct.
#
# Let's re-test with Example 3: n = 3, k = 2, invocations = [[1,2],[0,1],[2,0]]
# suspicious_set = {0, 1, 2}
#
# Check `can_remove_any`:
# m = 0: rev_adj[0] = [2]. Caller 2 IS in suspicious_set.
# m = 1: rev_adj[1] = [0]. Caller 0 IS in suspicious_set.
# m = 2: rev_adj[2] = [1]. Caller 1 IS in suspicious_set.
#
# `can_remove_any` remains True.
#
# Since `can_remove_any` is True, return methods NOT in `suspicious_set`.
# Methods not in {0, 1, 2} is empty.
# Output: []. Correct.
#
# This revised algorithm seems to capture the problem statement correctly.
#
# Time Complexity:
# 1. Building graphs: O(N + V), where V is the number of invocations.
# 2. BFS to find suspicious_set: O(N + V) in the worst case (dense graph).
# 3. Checking `can_remove_any`: We iterate through each method in `suspicious_set` (at most N). For each method, we iterate through its callers. The total number of calls considered across all methods in `suspicious_set` is the sum of in-degrees of methods in `suspicious_set`. In the worst case, this is O(V). So, this step is O(N + V).
# 4. Constructing result: O(N).
#
# Overall Time Complexity: O(N + V)
# Space Complexity:
# 1. Adjacency lists (`adj`, `rev_adj`): O(N + V).
# 2. `suspicious_set` and `queue` for BFS: O(N).
#
# Overall Space Complexity: O(N + V).
#
# Constraints: n <= 10^5, invocations.length <= 2 * 10^5. N+V is up to ~3*10^5, which is fine.

# Implementation details:
# Use `collections.defaultdict(list)` for adjacency lists.
# Use `collections.deque` for BFS queue.
# Use a `set` for `suspicious_set` for O(1) average time lookups.

import collections

class Solution:
    def removeMethods(self, n: int, k: int, invocations: list[list[int]]) -> list[int]:
        # Build forward adjacency list (adj[u] = list of methods v that u invokes)
        adj = collections.defaultdict(list)
        # Build reverse adjacency list (rev_adj[v] = list of methods u that invoke v)
        rev_adj = collections.defaultdict(list)

        for u, v in invocations:
            adj[u].append(v)
            rev_adj[v].append(u)

        # Step 2: Use BFS from k to find all suspicious methods
        suspicious_set = set()
        queue = collections.deque([k])
        suspicious_set.add(k)

        while queue:
            current_method = queue.popleft()

            # Explore methods invoked by the current_method
            for neighbor in adj[current_method]:
                if neighbor not in suspicious_set:
                    suspicious_set.add(neighbor)
                    queue.append(neighbor)

        # Step 3: Check if any suspicious method is invoked by a non-suspicious method.
        # If such a case exists, then NO suspicious methods can be removed according to the problem statement.
        can_remove_all_suspicious = True
        for suspicious_method in suspicious_set:
            # Check all callers of this suspicious method
            for caller in rev_adj[suspicious_method]:
                # If a caller is NOT suspicious, then this suspicious_method cannot be removed.
                # According to the problem statement, if any part of the suspicious group cannot be removed,
                # then none of the suspicious methods can be removed.
                if caller not in suspicious_set:
                    can_remove_all_suspicious = False
                    break # No need to check other callers for this suspicious_method
            if not can_remove_all_suspicious:
                break # No need to check other suspicious_methods

        # Step 4 & 5: Construct the result based on the check.
        if not can_remove_all_suspicious:
            # If it's not possible to remove all suspicious methods, return all methods.
            return list(range(n))
        else:
            # If all suspicious methods can be removed (i.e., they are not invoked by any non-suspicious method),
            # then return the methods that are NOT suspicious.
            remaining_methods = []
            for i in range(n):
                if i not in suspicious_set:
                    remaining_methods.append(i)
            return remaining_methods

```

File: solutions/2026-08-05/java/2026-08-05-remove-methods-from-project.java
```java
import java.util.*;

/**
 * Problem: Remove Methods From Project
 * Link: https://leetcode.com/problems/remove-methods-from-project/
 *
 * Approach:
 * This problem can be modeled as a directed graph where methods are nodes and invocations are edges.
 * We need to identify a set of suspicious methods. A method is suspicious if it is method 'k' or
 * if it directly or indirectly invokes method 'k'.
 * After identifying all suspicious methods, we need to check if any non-suspicious method invokes
 * a suspicious method. If such an invocation exists, we cannot remove the suspicious methods, and
 * all methods remain. Otherwise, we remove all suspicious methods.
 *
 * Steps:
 * 1. Build the adjacency list for the graph representing method invocations.
 * 2. Identify all suspicious methods using Depth First Search (DFS) or Breadth First Search (BFS)
 *    starting from method 'k'. Any node reachable from 'k' (including 'k' itself) is suspicious.
 *    Alternatively, we can think of it as finding all nodes that can reach 'k' in the reversed graph,
 *    but since 'a' invoking 'b' means an edge a -> b, then if 'a' invokes 'k', 'a' is suspicious.
 *    So, we need to find all nodes that can reach 'k' in the original graph.
 *    A simpler way is to find all nodes that are part of any path starting from 'k' or that lead to 'k'.
 *    Let's rephrase: method 'x' is suspicious if 'x' == k OR if 'x' invokes 'y' and 'y' is suspicious.
 *    This implies finding all nodes that are 'k' or can reach 'k'. This is equivalent to finding all nodes
 *    that are 'k' or that 'k' can reach in the reversed graph.
 *    Let's simplify: A method is suspicious if it is 'k' or invoked by 'k' (directly or indirectly).
 *    This means we need to find all nodes reachable from 'k' in the original graph.
 *    We also need to consider that if a non-suspicious method invokes a suspicious method, we can't remove.
 *
 *    Revised Approach:
 *    1. Build the graph where an edge `u -> v` means method `u` invokes method `v`.
 *    2. Find all methods that are "suspicious". A method `m` is suspicious if:
 *       a. `m == k`
 *       b. `m` directly or indirectly invokes `k`. This means there is a path from `m` to `k`.
 *       We can find all such methods by performing a DFS/BFS starting from `k` on the *reversed* graph.
 *       Let's call this set `S_invoking_k`.
 *       Also, any method invoked by a suspicious method (directly or indirectly) is also suspicious.
 *       This implies finding all methods reachable *from* `k` in the original graph. Let's call this set `S_reachable_from_k`.
 *       The set of all suspicious methods is `S = {k} U S_invoking_k U S_reachable_from_k`.
 *
 *    Let's simplify the definition of suspicious methods again.
 *    "Method k, along with any method invoked by it, either directly or indirectly, are considered suspicious"
 *    This means we need to find all methods reachable from `k` in the original graph. Let's call this set `SuspiciousSet`.
 *
 *    3. Now, we need to check the condition for removal: "A group of methods can only be removed if no method outside the group invokes any methods within it."
 *       This means if there is any edge `u -> v` where `u` is NOT in `SuspiciousSet` and `v` IS in `SuspiciousSet`, then we cannot remove the `SuspiciousSet`.
 *       If no such edge exists, we can remove all methods in `SuspiciousSet`.
 *
 *    Implementation Details:
 *    - Adjacency list for the graph: `adj[u]` stores methods invoked by `u`.
 *    - Reversed adjacency list: `revAdj[v]` stores methods that invoke `v`.
 *    - `suspicious` set: To store all methods that are suspicious.
 *    - `visited` array for DFS/BFS.
 *
 *    Let's refine the suspicious set logic:
 *    "Method k, along with any method invoked by it, either directly or indirectly, are considered suspicious"
 *    This means if we can reach method `m` from `k` (k -> ... -> m), then `m` is suspicious.
 *    So, we perform a DFS/BFS starting from `k` on the original graph to find all reachable nodes.
 *    Let `reachable_from_k` be the set of nodes reachable from `k`.
 *    The set of suspicious methods `SuspiciousSet` is `reachable_from_k`.
 *
 *    Now check removal condition:
 *    Iterate through all invocations `[a, b]`.
 *    If `b` is in `SuspiciousSet` AND `a` is NOT in `SuspiciousSet`, then we cannot remove.
 *
 *    Example 1: n = 4, k = 1, invocations = [[1,2],[0,1],[3,2]]
 *    adj: 0:[1], 1:[2], 3:[2]
 *    revAdj: 1:[0], 2:[1,3]
 *    Start DFS from k=1:
 *    Reachable from 1: {1, 2} (since 1 invokes 2, and nothing else is invoked by 2).
 *    SuspiciousSet = {1, 2}
 *    Check removal:
 *    [1, 2]: 1 is suspicious, 2 is suspicious. OK.
 *    [0, 1]: 1 is suspicious. Is 0 NOT suspicious? Yes. So, non-suspicious (0) invokes suspicious (1).
 *            Cannot remove. Return all methods: [0, 1, 2, 3].
 *
 *    Example 2: n = 5, k = 0, invocations = [[1,2],[0,2],[0,1],[3,4]]
 *    adj: 0:[2,1], 1:[2], 3:[4]
 *    revAdj: 1:[0], 2:[1,0], 4:[3]
 *    Start DFS from k=0:
 *    Reachable from 0: {0, 1, 2} (0->1, 0->2, 1->2).
 *    SuspiciousSet = {0, 1, 2}
 *    Check removal:
 *    [1, 2]: 1 is suspicious, 2 is suspicious. OK.
 *    [0, 2]: 0 is suspicious, 2 is suspicious. OK.
 *    [0, 1]: 0 is suspicious, 1 is suspicious. OK.
 *    [3, 4]: 3 is NOT suspicious. 4 is NOT suspicious. OK.
 *    No non-suspicious method invokes a suspicious method. Remove {0, 1, 2}.
 *    Remaining methods: {3, 4}. Return [3, 4].
 *
 *    Example 3: n = 3, k = 2, invocations = [[1,2],[0,1],[2,0]]
 *    adj: 0:[1], 1:[2], 2:[0]
 *    revAdj: 0:[2], 1:[0], 2:[1]
 *    Start DFS from k=2:
 *    Reachable from 2: {2, 0, 1} (2->0, 0->1, 1->2).
 *    SuspiciousSet = {0, 1, 2}
 *    Check removal:
 *    All methods are suspicious. There are no non-suspicious methods. The condition "no method outside the group invokes any methods within it" is vacuously true.
 *    Remove {0, 1, 2}. Remaining methods: []. Return [].
 *
 *    Complexity:
 *    - Building graph: O(N + E) where N is number of methods, E is number of invocations.
 *    - Finding suspicious methods (DFS/BFS from k): O(N + E).
 *    - Checking removal condition: O(E).
 *    - Collecting remaining methods: O(N).
 *    Overall Time Complexity: O(N + E)
 *    Space Complexity: O(N + E) for adjacency lists and visited set.
 */
class Solution {
    private List<List<Integer>> adj; // Adjacency list: adj[u] contains methods invoked by u
    private List<List<Integer>> revAdj; // Reversed adjacency list: revAdj[v] contains methods that invoke v
    private Set<Integer> suspiciousMethods; // Stores all methods deemed suspicious
    private boolean canRemove; // Flag to indicate if all suspicious methods can be removed

    public List<Integer> removeMethods(int n, int k, int[][] invocations) {
        // Initialize adjacency lists
        adj = new ArrayList<>(n);
        revAdj = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
            revAdj.add(new ArrayList<>());
        }

        // Build the graph and the reversed graph from invocations
        for (int[] invocation : invocations) {
            int u = invocation[0]; // Method u invokes method v
            int v = invocation[1];
            adj.get(u).add(v);
            revAdj.get(v).add(u);
        }

        // 1. Identify all suspicious methods.
        // A method is suspicious if it is k OR it is invoked by k directly or indirectly.
        // This means all nodes reachable FROM k in the original graph are suspicious.
        suspiciousMethods = new HashSet<>();
        boolean[] visitedForSuspicious = new boolean[n];
        dfsFindSuspicious(k, visitedForSuspicious);

        // 2. Check if any non-suspicious method invokes a suspicious method.
        // If such an invocation exists, we cannot remove the suspicious methods.
        canRemove = true; // Assume we can remove until proven otherwise
        for (int u = 0; u < n; u++) {
            // If method 'u' is NOT suspicious
            if (!suspiciousMethods.contains(u)) {
                // Check all methods 'v' that 'u' invokes
                for (int v : adj.get(u)) {
                    // If 'v' IS suspicious, then a non-suspicious method 'u' invokes a suspicious method 'v'
                    if (suspiciousMethods.contains(v)) {
                        canRemove = false; // Cannot remove
                        break; // No need to check further for this 'u'
                    }
                }
            }
            if (!canRemove) {
                break; // No need to check other non-suspicious methods
            }
        }

        // 3. Prepare the result.
        List<Integer> remainingMethods = new ArrayList<>();
        if (!canRemove) {
            // If not all suspicious methods can be removed, return all methods.
            for (int i = 0; i < n; i++) {
                remainingMethods.add(i);
            }
        } else {
            // If all suspicious methods can be removed, return only the non-suspicious methods.
            for (int i = 0; i < n; i++) {
                if (!suspiciousMethods.contains(i)) {
                    remainingMethods.add(i);
                }
            }
        }

        return remainingMethods;
    }

    /**
     * Performs DFS starting from a given method to find all methods reachable from it.
     * These reachable methods are considered suspicious.
     * @param currentMethod The current method being visited.
     * @param visited Array to keep track of visited nodes during DFS to avoid cycles and redundant work.
     */
    private void dfsFindSuspicious(int currentMethod, boolean[] visited) {
        // Mark the current method as visited and add it to the set of suspicious methods.
        visited[currentMethod] = true;
        suspiciousMethods.add(currentMethod);

        // Explore all methods that the current method invokes.
        for (int neighbor : adj.get(currentMethod)) {
            // If a neighbor has not been visited yet, recursively call DFS on it.
            if (!visited[neighbor]) {
                dfsFindSuspicious(neighbor, visited);
            }
        }
    }
}
```